Bluetooth: controller: Fix advertising initialization on reset

andrzej-kaczmarek opened the pull request:
This fixes advertising initialization on reset:
- all sets are disabled prior to cleanup
- all sets are removed after reset
- set #0 is initialized with default legacy PDUs

andrzej-kaczmarek committed:
Bluetooth: controller: Disable advertising before cleanup on reset

It seems more reasonable to first disable all advertising sets before
final cleanup.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Remove all advertising sets on reset

We should remove all advertising sets when resetting controller. It's
enough to just clear is_created flag for each set.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Initialize PDUs in set #0 on reset

Set #0 is used for legacy advertising when legacy HCI commands are used
and in such case it's possible to enable advertising without any prior
configuration. This means we should better always have PDUs in set #0
initialized with default empty data.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1319,6 +1321,10 @@ int ull_adv_reset(void)
 {
 	uint8_t handle;
 
+	for (handle = 0U; handle < BT_CTLR_ADV_SET; handle++) {
```
accepted this change, but reset of Periodic Advertising is missing and can be added in another PR.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1345,11 +1351,9 @@ int ull_adv_reset(void)
-	for (handle = 0U; handle < BT_CTLR_ADV_SET; handle++) {
-		(void)disable(handle);
-	}
+	ll_adv_aux_set_clear();
```
```suggestion
```

Instead add `ll_adv[handle].is_created = 0U;` inside here: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/ull_adv.c#L1363

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1569,6 +1573,13 @@ static int init_reset(void)
 		lll_adv_data_init(&ll_adv[handle].lll.scan_rsp);
 	}
 
+	/* Make sure that set #0 is initialized with empty legacy PDUs. This is
+	 * especially important if legacy HCI interface is used for advertising
+	 * because it allows to enable advertising without any configuration,
+	 * thus we need to have PDUs already initialized.
+	 */
+	init_legacy_pdu(&ll_adv[0]);
+
```
```suggestion
	
```

Instead do the common initialization of bit fields and len here: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c#L150
And, set the `type` after the call to `lll_adv_data_init` here: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/ull_adv.c#L1568

andrzej-kaczmarek replied:
@cvinayak This would do a lot of redundant initialization. `lll_adv_data_init` is also used for aux and sync which initialize PDU from scratch anyway. Also we don't really need to initialize anything other that set #0, because those sets will be initialized on create. The only reason we need to initialize set #0 is because legacy advertising can be started by legacy HCI command without any prior configuration.

So I'd keep this initialization separately. I can just rework it a bit into smth like `lll_adv_pdu_init` and call it for `&ll_adv[0].lll.adv_data` and `&ll_adv[0].lll.scan_rsp`. Alternatively, initialization can be postponed until we switch to legacy HCI commands when 1st legacy command is used.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1940,3 +1951,26 @@ static inline uint8_t disable(uint8_t handle)
 
 	return 0;
 }
+
+static void init_legacy_pdu(struct ll_adv_set *adv)
+{
+	struct pdu_adv *pdu;
+
+	/* Initialize default ADV_IND */
+	pdu = lll_adv_data_peek(&adv->lll);
+	pdu->type = PDU_ADV_TYPE_SCAN_RSP;
+	pdu->rfu = 0;
+	pdu->chan_sel = 0;
+	pdu->tx_addr = 0;
+	pdu->rx_addr = 0;
+	pdu->len = BDADDR_SIZE;
+
+	/* Initialize default SCAN_RSP */
+	pdu = lll_adv_scan_rsp_peek(&adv->lll);
+	pdu->type = PDU_ADV_TYPE_SCAN_RSP;
+	pdu->rfu = 0;
+	pdu->chan_sel = 0;
+	pdu->tx_addr = 0;
+	pdu->rx_addr = 0;
+	pdu->len = BDADDR_SIZE;
+}
```
```suggestion

```

andrzej-kaczmarek committed:
Bluetooth: controller: Improve readability

Create shortcuts to set and lll to make code easier to read.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Fix reset with ext adv enabled

We need to make sure aux is cleared on reset, otherwise we may end up
with legacy advertising with aux.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1569,6 +1578,14 @@ static int init_reset(void)
 		lll_adv_data_init(&ll_adv[handle].lll.scan_rsp);
 	}
 
+	/* Make sure that set #0 is initialized with empty legacy PDUs. This is
+	 * especially important if legacy HCI interface is used for advertising
+	 * because it allows to enable advertising without any configuration,
+	 * thus we need to have PDUs already initialized.
+	 */
+	init_pdu(lll_adv_data_peek(&ll_adv[0].lll), PDU_ADV_TYPE_ADV_IND);
+	init_pdu(lll_adv_scan_rsp_peek(&ll_adv[0].lll), PDU_ADV_TYPE_SCAN_RSP);
+
```
This change is redundant (added CPU use on HCI reset) in Zephyr BT Host usage, as advertising parameters are always set after HCI reset before enabling advertising. Most third party hosts too do that same. (Also, currently there is not default advertising parameters initialized that would allow proper advertising enable without parameters being set by host).

andrzej-kaczmarek replied:
I will also handle default parameters on reset and put it inside #ifdef so it's not used with Zephyr host. Regardless of what hosts do, some qualification test cases do not set parameters explicitly and will fail randomly if we don't initialize parameters properly.

carlescufi commented:
@andrzej-kaczmarek please rebase

andrzej-kaczmarek commented:
@carlescufi rebased

carlescufi commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1600,6 +1612,18 @@ static int init_reset(void)
+	 * because it allows to enable advertising without any configuration,
+	 * thus we need to have PDUs already initialized.
+	 *
+	 * Note: not needed if Zephyr host is used since it always configures
```
Should we really rely on the Host doing this? Why not do this unconditionally?
cc @joerchan

andrzej-kaczmarek replied:
I reverted this back to original version (unconditional init). This does not really do any harm and has no real impact on power consumption but just makes code cleaner. It can be optimized later especially if "trusted host" setting is introduced as suggested before.

joerchan commented on include/bluetooth/hci.h:
```diff
@@ -815,6 +815,15 @@ struct bt_hci_cp_le_set_random_address {
+#define BT_LE_ADV_INTERVAL_MAX                  0x4000
+#define BT_LE_ADV_INTERVAL_DEFAULT              0x0800
+
+#define BT_LE_ADV_CHAN_MAP_CH37                 0x01
+#define BT_LE_ADV_CHAN_MAP_CH38                 0x02
+#define BT_LE_ADV_CHAN_MAP_CH39                 0x04
```
For consistency I would name them CHAN_37/38/39 (bluetooth.h has this).

andrzej-kaczmarek replied:
fixed

carlescufi commented:
@andrzej-kaczmarek could you follow-up on this one?
